### PR TITLE
Add copy methods to hiopMatrix interface

### PR DESCRIPTION
**Pull request by ashermancinelli**
As per #25 adds methods to interface. What other methods may we need to add to the interface? How can we reconcile the `addSubDiagonal` implementations?

### CODE REVIEW

**Comment by cnpetra** [comment]
I am working on addSubDiagonal in a different branch, let's leave it out for now.

File: src/LinAlg/hiopMatrix.hpp [hunk 1]
```diff
@@ -84,6 +84,16 @@ class hiopMatrix
   virtual hiopMatrix* alloc_clone() const=0;
   virtual hiopMatrix* new_copy() const=0;
 
+  virtual void copyFrom(const hiopMatrix& matrix) = 0;
+  virtual void copyFrom(const double* buffer) = 0;
+
+  /* copies 'src' into this as a block starting at (i_block_start,j_block_start) */
+  void copyBlockFromMatrix(const long i_block_start, const long j_block_start, const hiopMatrix& src) = 0;
```
Comment (cnpetra) [inline src/LinAlg/hiopMatrix.hpp:91]: this method makes sense only to particular matrices (dense and maybe mds), and I would not make it part of the interface.

File: src/LinAlg/hiopMatrix.hpp [hunk 2]
```diff
@@ -84,6 +84,16 @@ class hiopMatrix
   virtual hiopMatrix* alloc_clone() const=0;
   virtual hiopMatrix* new_copy() const=0;
 
+  virtual void copyFrom(const hiopMatrix& matrix) = 0;
+  virtual void copyFrom(const double* buffer) = 0;
```
Comment (cnpetra) [inline src/LinAlg/hiopMatrix.hpp:88]: this method does not make sense for sparse of mds matrices (specific to dense matrices in other words). I would not make it part of the interface.

File: src/LinAlg/hiopMatrix.hpp [hunk 3]
```diff
@@ -84,6 +84,16 @@ class hiopMatrix
...
+
+  /* copies 'src' into this as a block starting at (i_block_start,j_block_start) */
+  void copyBlockFromMatrix(const long i_block_start, const long j_block_start, const hiopMatrix& src) = 0;
+
+  /* overwrites 'this' with 'src''s block that starts at (i_src_block_start,j_src_block_start) 
+   * and has dimensions of 'this' */
+  void copyFromMatrixBlock(const hiopMatrix& src, const int i_src_block_start, const int j_src_block_start) = 0;
```
Comment (cnpetra) [inline src/LinAlg/hiopMatrix.hpp:95]: this method is needed only for dense and can be efficiently implemented only dense matrices. I would not make it part of the interface.

**Comment by cnpetra** [comment]
hiopVector may be in need of a lifting a couple of methods, see #23.

**Comment by ashermancinelli** [comment]
You may PR my branch for the sake of cleanliness, or you can send me a diff that I can apply manually. However you would like to proceed is well with me.

**Comment by cnpetra** [comment]
better to not lift the `copyXXX` methods as previously discussed

`addSubDiagonal` is addressed in PR#42 and will move the discussion there

will be closing this (without merge), but feel free to reopen if I missed anything.

**Pull request closed without merging**